runtime/src/deterministic.rs

```diff
@@ -690,15 +783,36 @@ impl Context {
         Self {
             name: String::new(),
             spawned: false,
-            executor,
+            executor: Arc::downgrade(&executor),
+            owner: Some(executor.clone()),
             network: Arc::new(network),
             storage: self.storage,
+            metrics,
+            auditor,
             children: Arc::new(Mutex::new(Vec::new())),
         }
     }
 
     pub fn auditor(&self) -> &Auditor {
-        &self.executor.auditor
+        &self.auditor
+    }
+
+    // Upgrade executor Weak reference; panic if unavailable
+    fn executor(&self) -> Arc<Executor> {
+        self.executor
+            .upgrade()
+            .expect("executor dropped while context still in use")
+    }
+
+    // Make a detached context without a strong Executor owner
+    fn detached(mut self) -> Self {
+        self.owner = None;
```

Comment:
We should remove `.owner()` from `Context`. It makes it much more confusing to follow.

Reply:
When doing so, we'll need to be careful not to break `recover` (as it relies on upgrading to a strong reference)?